samples: lwm2m_client: FOTA python script's (#10472)
nrfconnect/sdk-nrf: samples/nrf9160/lwm2m_client/fota.rst

+-------------+-------------+---------------------------------------------------------------------+

Advanced Firmware Update helper script
**************************************

For working with the `AVSystem Coiote server <Coiote Device Management_>`_, a helper script has been provided.
To automate the firmware update, you can use the script :file:`fota.py` that is available in the :file:`samples/nrf9160/lwm2m_client/scripts/` folder.
The :file:`fota.py` file supports ``update`` and ``upload`` commands for firmware updates.

The commands can be used in the following way:

.. parsed-literal::

   ./scripts/fota.py -id *device_id* -to *time_out* update *instance_id* *binary_type* *binary_src*
   ./scripts/fota.py upload *instance_id* *binary_name*

FOTA upload command
===================

The FOTA ``upload`` command allocates a resource ID and uploads the binary to Coiote Device Management server.
Currently, the script defines resource ID based on the given instance ID: ``lwm2m_client_fota_instance_<instance_id>``.

.. code-block:: console

   ./scripts/fota.py upload 0 app_update.bin

> The following example shows the allocation of a resource id ``lwm2m_client_fota_instance_0`` on the server side:

Reviewer: ```suggestion
The following example output shows the allocation of a resource id ``lwm2m_client_fota_instance_0`` on the server side:
```

.. code-block:: console

   [INFO] fota.py - Upload app_update.bin with instance 0 to Coiote
   [INFO] fota.py - Binary app_update.bin, Size 336032 (bytes)
   [INFO] coiote.py - Creating fota resource for binary app_update.bin with id lwm2m_client_fota_instance_0
   [INFO] fota.py - Allocated Resource id lwm2m_client_fota_instance_0 for instance 0

.. code-block:: console

   ./scripts/fota.py upload 1 mfw_nrf9160_update_from_1.3.4_to_1.3.4-FOTA-TEST.bin

> The following example shows the allocation of a resource id ``lwm2m_client_fota_instance_0`` on the server side:

Reviewer: ```suggestion
The following example output shows the allocation of a resource id ``lwm2m_client_fota_instance_0`` on the server side:
```

      .. code-block:: console

         # Setup phase
         export COIOTE_PASSWD='my-password'
         export COIOTE_USER='my-username'

   #. To use the :ref:`lwm2m_client` sample after updating the firmware, you must build the sample with the following overlays:

         * :file:`overlay-avsystems.conf`
         * :file:`overlay-lwm2m-1.1.conf`
         * :file:`overlay-fota_helper.conf`
         * :file:`overlay-adv-firmware.conf`

   #. Register your device with the Coiote Device management server.
   #. Flash the compiled sample using the erase flash option.
   #. Wait for the device registration to be complete.
   #. Open the :file:`src/prj.conf` file.
   #. Change :kconfig:option:`CONFIG_MCUBOOT_IMAGE_VERSION` to ``1.0.1`` and rebuild the sample.
   #. Update the application and modem firmware by using the :file:`/scripts/fota.py` script:

      .. code-block:: console

         ./scripts/fota.py -id urn:imei:351358811331351 update 0 file app_update.bin update 1 file mfw_nrf9160_update_from_1.3.4_to_1.3.4-FOTA-TEST.bin

>    #. Following is an example output of the command:

Reviewer: This is not a step or a user action and hence can be part of the previous step.

Author: Ok I will fix this.

         [INFO] coiote.py - Device is Queuemode Coiote have to wait next Registration Update
         [INFO] fota.py - Download Url Write done
         [INFO] fota.py - Enable Observation Advanced Firmware Update
         [INFO] fota.py - Downloading instance: 0
         [INFO] fota.py - Downloading instance: 1
         [INFO] fota.py - Download ready instance: 0
         [INFO] fota.py - Download ready instance: 1
         [INFO] coiote.py - Device is Queuemode Coiote have to wait next Registration Update
         [INFO] fota.py - Update started instance: 0
         [INFO] fota.py - Update started instance: 1
         [INFO] fota.py - Firmware Update Successfully instance: 0
         [INFO] fota.py - From:1.0.0-0 to 1.0.1-0
         [INFO] fota.py - Firmware Update Successfully instance: 1
         [INFO] fota.py - From:mfw_nrf9160_1.3.4 to mfw_nrf9160_1.3.4-FOTA-TEST
         [INFO] fota.py - Firmware update process finished
         [INFO] fota.py - Delete Observation Advanced Firmware Update

   #. Update the modem firmware back to the original released version:

      .. code-block:: console

>          ./scripts/fota.py -id urn:imei:351358811331351 update 1 file mfw_nrf9160_update_from_1.3.4-FOTA-TEST_to_1.3.4.bin

Reviewer: Don't we need to change the version format to 1.x.x as in lines 203 and 204?

Author: I updated this and next command example to use 1.x.x

>          ./scripts/fota.py -id urn:imei:351358811331351 update 1 file mfw_nrf9160_update_from_1.3.4-FOTA-TEST_to_1.x.x.bin

Reviewer: ```suggestion
         ./scripts/fota.py -id urn:imei:351358811331351 update 1 file mfw_nrf9160_update_from_1.x.x-FOTA-TEST_to_1.x.x.bin
```

>    #. Following is an example output of the command:

Reviewer: This is not a step or a user action and hence can be part of the previous step.

Author: I can fix this also

| Resource ID | Name              | Operation | Description                                        |
+=============+===================+===========+====================================================+
| 14          | Component name    | Read      | Name of the component handled by this instance of  |
|             |                   |           | the Advanced Firmware Update object.               |
+-------------+-------------------+-----------+----------------------------------------------------+
| 15          | Current version   | Read      | Version number of the package that is currently    |
|             |                   |           | installed and running for the component handled    |
|             |                   |           | by this instance of the Advanced Firmware Update   |
|             |                   |           | object.                                            |
+-------------+-------------------+-----------+----------------------------------------------------+
| 16          | Linked instance   | Read      | When multiple instances of the Advanced Firmware   |

> |             |                   |           | Update object are in the downloaded state, this    |

Reviewer: ```suggestion
|             |                   |           | Update object are in the ``downloaded`` state, this    |
```

.. code-block:: console

   [INFO] fota.py - Upload mfw_nrf9160_update_from_1.3.4_to_1.3.4-FOTA-TEST.bin with instance 1 to Coiote
   [INFO] fota.py - Binary mfw_nrf9160_update_from_1.3.4_to_1.3.4-FOTA-TEST.bin, Size 14472 (bytes)
   [INFO] coiote.py - Creating fota resource for binary mfw_nrf9160_update_from_1.3.4_to_1.3.4-FOTA-TEST.bin with id lwm2m_client_fota_instance_1
   [INFO] fota.py - Allocated Resource id lwm2m_client_fota_instance_1 for instance 1

Fota update command
-------------------

For running the full FOTA process, firmware can be given as a resource ID that already exists on the server side, or as a binary file that will be uploaded automatically.

To use a binary name:

+-----------------+--------------------------------+---------+--------------------------------------+
| Python file     | Options                        | Command | Arguments with binary file name      |
+=================+================================+=========+======================================+
| :file:`fota.py` | -id <device_id> -to <time_out> | update  | <instance_id> file  <binary_name>    |

> +-----------------+--------------------------------+---------+--------------------------------------+

Author: Hi @umapraseeda . I have not not used command format so thats why I was using table

.. code-block:: console

   ./scripts/fota.py -id urn:imei:351358811331351 update 1 file mfw_nrf9160_update_from_1.3.4_to_1.3.4-FOTA-TEST.bin

To use existing resource IDs:

+-----------------+-----------------+----------+------------------------------------------+
| Python file     | Options         | Command  | Arguments with Resource ID               |
+=================+=================+==========+==========================================+
| :file:`fota.py` | -id <device_id> | update   | <instance_id> resource  <resource_id>    |
+-----------------+-----------------+----------+------------------------------------------+

Following is an example of uploading a binary and updating a modem by referring to the uploaded resource ID:

.. code-block:: console

   ./scripts/fota.py upload 1 mfw_nrf9160_update_from_1.3.4-FOTA-TEST_to_1.3.4.bin
   ./scripts/fota.py -id urn:imei:351358811331351 update 1 resource lwm2m_client_fota_instance_1

Testing advanced FOTA
=====================

Complete the following steps to test the advanced Fota firmware update with the lwm2m client sample and the :file:`/scripts/fota.py` file.

>    #. Download the released modem version 1.3.4 zip file from `nRF9160 DK Downloads`_.

Author: Yes It could be latest but it will then effect how to select fota delta test binaries.

>    #. Update the modem firmware using the nRF Programmer app of `nRF Connect for Desktop`_.

Reviewer 2: Is it for the nRF Programmer app?
If so, 
The nRF nRF Connect for Desktop link (Website page, which is already mentioned here) has a description about nRF Programmer app.

The other option could be to link to the infocenter page: https://infocenter.nordicsemi.com/index.jsp?topic=%2Fug_nc_programmer%2FUG%2Fnrf_connect_programmer%2Fncp_introduction.html 

I think it would be better to link to the infocenter page. If so, I can make this change in my commit.

>    #. Update the modem firmware using the nRF Programmer app of `nRF Connect for Desktop`_.

Reviewer: No, I meant to link to the instructions for updating the modem firmware on nRF9160-based devices.

   #. Copy the following binaries from the zip file to the folder :file:`/nrf/samples/nrf9160/lwm2m_client`

         * mfw_nrf9160_update_from_1.3.4_to_1.3.4-FOTA-TEST.bin

>          * mfw_nrf9160_update_from_1.3.4-FOTA-TEST_to_1.3.4.bin

Author: If we use download latest may be we should change 1.3.4 -> 1.x.x or something?

   #. To set up the script, you must set your username and password of the AVSystem's Coiote Device Management server as environment variables.

      .. code-block:: console

         # Setup phase
         export COIOTE_PASSWD='my-password'
         export COIOTE_USER='my-username'

   #. To use the :ref:`lwm2m_client` sample in updated firmware, you must build the client with the following overlays:

         * :file:`overlay-avsystems.conf`
         * :file:`overlay-lwm2m-1.1.conf`
         * :file:`overlay-fota_helper.conf`
         * :file:`overlay-adv-firmware.conf`

   #. Register device with the Coiote Device management server.

>    #. Flash compiled sample with an erase flash option to the device.

Author: Yes

>    #. Flash compiled sample with an erase flash option to the device.

Reviewer 2: I have kept a generic statement because the user can erase and flash using command line also.
Please check.

Library support
***************

Advanced Firmware update is part of the :ref:`lib_lwm2m_client_utils` library and is designed to support updating of modem and application binary updates.

Configuration option
====================

Enable the following Kconfig options to use advanced FOTA:

.. code-block:: console

   CONFIG_LWM2M_CLIENT_UTILS=y
   CONFIG_LWM2M_CLIENT_UTILS_ADV_FIRMWARE_UPDATE_OBJ_SUPPORT=y
   CONFIG_LWM2M_CLIENT_UTILS_FIRMWARE_UPDATE_OBJ_SUPPORT=y
   CONFIG_LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT=n
   CONFIG_FOTA_CLIENT_AUTOSCHEDULE_UPDATE=n

The :file:`overlay-adv-firmware.conf` configuration file includes these options.

Initializing
============

The Advanced Firmware Update object shares the same API with the standard LwM2M Firmware Update object as implemented in :ref:`lib_lwm2m_client_utils`.
The Advanced Firmware Update functionality is implemented in :file:`lwm2m_firmware.c` and :file:`lwm2m_adv_firmware.c`.

The Advanced Firmware Update object is initialized by calling the :c:func:`lwm2m_init_firmware` function and by confirming the currently booted application image as valid by calling the :c:func:`lwm2m_init_image` function.

> The result of validating the image marks the FOTA process as completed and is reported to the server.

Reviewer: ```suggestion
The validation of the image marks the FOTA process as complete  and the result  is reported to the server.
```

====================================

By default, the Advanced Firmware Update object supports two instances that are configured in the following way:

+-------------+-------------+---------------------------------------------------------------------+
| Instance ID | Owner       | DFU types                                                           |
+=============+=============+=====================================================================+
| 0           | Application | DFU_TARGET_IMAGE_TYPE_MCUBOOT                                       |
+-------------+-------------+---------------------------------------------------------------------+
| 1           | Modem       | DFU_TARGET_IMAGE_TYPE_MODEM_DELTA, DFU_TARGET_IMAGE_TYPE_FULL_MODEM |
+-------------+-------------+---------------------------------------------------------------------+

Advanced Firmware Update helper script
--------------------------------------

For working with the `AVSystem Coiote server <Coiote Device Management_>`_, a helper script has been provided.
To automate the firmware update, you can use the script :file:`fota.py` that is available in the :file:`samples/nrf9160/lwm2m_client/scripts/` folder.

> For firmware updates, the :file:`/fota.py` file supports ``update`` and ``upload`` commands.

Reviewer: ```suggestion
For firmware updates, the :file:`fota.py` file supports ``update`` and ``upload`` commands.
```

> +-----------------+-------------------------------+-------------+-------------------------------------------------+

Author: @umapraseeda and @divipillai 
Shuold we also change this table to command or keep it?

Reviewer: Good point. Yes, we can. Added a suggestion.

> | Python file     | Options                       | Command     | Arguments                                       |
> +=================+===============================+=============+=================================================+
> | :file:`fota.py` | -id <device_id> -to <time_out>| ``update``  | ``Instance_id``, ``binary_type``, ``binary_src``|
> +-----------------+-------------------------------+-------------+-------------------------------------------------+
> | :file:`fota.py` | None                          | ``upload``  | ``Instance_id``, ``binary_name``                |
> +-----------------+-------------------------------+-------------+-------------------------------------------------+

Reviewer: ```suggestion
The commands can be used in the following way:

.. parsed-literal::

   ./scripts/fota.py -id *device_id* -to *time_out* update *instance_id* *binary_type* *binary_src*
   ./scripts/fota.py upload *instance_id* *binary_name*
  
```

   [INFO] fota.py - Binary app_update.bin, Size 336032 (bytes)
   [INFO] coiote.py - Creating fota resource for binary app_update.bin with id lwm2m_client_fota_instance_0
   [INFO] fota.py - Allocated Resource id lwm2m_client_fota_instance_0 for instance 0

.. code-block:: console

   ./scripts/fota.py upload 1 mfw_nrf9160_update_from_1.3.4_to_1.3.4-FOTA-TEST.bin

The following example shows the allocation of a resource id ``lwm2m_client_fota_instance_0`` on the server side:

.. code-block:: console

   [INFO] fota.py - Upload mfw_nrf9160_update_from_1.3.4_to_1.3.4-FOTA-TEST.bin with instance 1 to Coiote
   [INFO] fota.py - Binary mfw_nrf9160_update_from_1.3.4_to_1.3.4-FOTA-TEST.bin, Size 14472 (bytes)
   [INFO] coiote.py - Creating fota resource for binary mfw_nrf9160_update_from_1.3.4_to_1.3.4-FOTA-TEST.bin with id lwm2m_client_fota_instance_1
   [INFO] fota.py - Allocated Resource id lwm2m_client_fota_instance_1 for instance 1

FOTA update command
-------------------

For running the full FOTA process, the firmware can be given as a resource ID that already exists on the server side, or as a binary file that will be uploaded automatically.

> To use a binary name:

Reviewer: ```suggestion
To use a binary name, run the script with the following parameters:
```

> +-----------------+--------------------------------+---------+--------------------------------------+
> | Python file     | Options                        | Command | Arguments with binary file name      |
> +=================+================================+=========+======================================+
> | :file:`fota.py` | -id <device_id> -to <time_out> | update  | <instance_id> file  <binary_name>    |
> +-----------------+--------------------------------+---------+--------------------------------------+

Reviewer: ```suggestion
.. parsed-literal::

   ./scripts/fota.py -id *device_id* -to *timeout* update *instance_id* file *binary_name*
```

When binary type is ``file``, script tries to discover given binary name from the :file:`samples/nrf9160/lwm2m_client/` file, or the :file:`samples/nrf9160/lwm2m_client/build/zephyr`.
The ``update`` command uploads and generates binary resource IDs automatically when a binary file is used.
The update default task timeout is 800 seconds.

Following is an example of updating a modem instance by giving a binary file:

.. code-block:: console

   ./scripts/fota.py -id urn:imei:351358811331351 update 1 file mfw_nrf9160_update_from_1.3.4_to_1.3.4-FOTA-TEST.bin

> To use existing resource IDs:

Reviewer: ```suggestion
To use the existing resource IDs, run the script with the following parameters:

```

> +-----------------+-----------------+----------+------------------------------------------+
> | Python file     | Options         | Command  | Arguments with Resource ID               |
> +=================+=================+==========+==========================================+
> | :file:`fota.py` | -id <device_id> | update   | <instance_id> resource  <resource_id>    |
> +-----------------+-----------------+----------+------------------------------------------+

Reviewer: ```suggestion
.. parsed-literal::

   ./scripts/fota.py -id *device_id* update *instance_id* resource *resource_id*
```

Following is an example of uploading a binary and updating a modem by referring to the uploaded resource ID:

.. code-block:: console

   ./scripts/fota.py upload 1 mfw_nrf9160_update_from_1.3.4-FOTA-TEST_to_1.3.4.bin
   ./scripts/fota.py -id urn:imei:351358811331351 update 1 resource lwm2m_client_fota_instance_1

Testing advanced FOTA
=====================

Complete the following steps to test the advanced FOTA firmware update with the lwM2M client sample and the :file:`/scripts/fota.py` file.

   #. Download the latest released modem zip file from `nRF9160 DK Downloads`_.
   #. Update the modem firmware using the nRF Programmer app of `nRF Connect for Desktop`_.

>    #. Copy the following binaries from the zip file to the folder :file:`/nrf/samples/nrf9160/lwm2m_client`

Reviewer: ```suggestion
   #. Copy the binaries with the following naming format from the zip file to the folder :file:`/nrf/samples/nrf9160/lwm2m_client`
```

>       * mfw_nrf9160_update_from_1.x.x_to_1.x.x-FOTA-TEST.bin
>       * mfw_nrf9160_update_from_1.x.x-FOTA-TEST_to_1.x.x.bin

Reviewer: ```suggestion
      * :file:`mfw_nrf9160_update_from_1.x.x_to_1.x.x-FOTA-TEST.bin`
      * :file:`mfw_nrf9160_update_from_1.x.x-FOTA-TEST_to_1.x.x.bin`
```

      where 1.x.x is the latest modem release version.

   #. To set up the script, you must set the username and password that you used in AVSystem's Coiote Device Management server as the environment variables.

      .. code-block:: console

         # Setup phase
         export COIOTE_PASSWD='my-password'
         export COIOTE_USER='my-username'

   #. To use the :ref:`lwm2m_client` sample after updating the firmware, you must build the sample with the following overlays:

         * :file:`overlay-avsystems.conf`
         * :file:`overlay-lwm2m-1.1.conf`
         * :file:`overlay-fota_helper.conf`
         * :file:`overlay-adv-firmware.conf`

   #. Register your device with the Coiote Device management server.

>    #. Flash compiled sample using the erase flash option.

Reviewer: ```suggestion
   #. Flash the compiled sample using the erase flash option.
```

>    #. Wait for the device to register.

Reviewer: ```suggestion
   #. Wait for the device registration to be complete.
```

   :local:
   :depth: 2

The advanced Firmware Update object (object id ``33629``) is an experimental object under development.
It allows to update multiple firmware packages from a device without deviating from the standard LwM2M Firmware Update object.
Advaced Firmware Update shares the first 13 resources with the OMA LwM2M Firmware Update object (object id ``5``) v1.1.
However, Zephyr only implements version 1.0 of the Firmaware Update object.

> For information about differences between these versions, see `LwM2M Firmware Update version 1.1 <LwM2M v1.1 specification_>`_.

Author: Here we should point to lwm2m version 1.2

Author: Firmware Update version 1.1 <LwM2M v1.2 specification_>`_.
